// hexo-component-inferno/lib/view/comment/giscus.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var _inferno = require("inferno");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
/**
 * Giscus comment JSX component.
 * @module view/comment/giscus
 */
var _require = require('inferno'),
  Component = _require.Component;
var _require2 = require('../../util/cache'),
  cacheComponent = _require2.cacheComponent;

/**
 * Giscus comment JSX component.
 *
 * @see https://giscus.app/
 * @example
 * <Giscus
 *     repo="usr/repo"
 *     repoId="X_xxxxxxxxxx"
 *     category="******"
 *     categoryId="XXX_xxxxxxxxxxxxxxxx"
 *     mapping="******"
 *     strict={0}
 *     reactionsEnabled={0}
 *     emitMetadata={0}
 *     inputPosition="******"
 *     theme="******"
 *     lang="******"
 *     lazy="******" />
 */
var Giscus = /*#__PURE__*/function (_Component) {
  function Giscus() {
    _classCallCheck(this, Giscus);
    return _callSuper(this, Giscus, arguments);
  }
  _inherits(Giscus, _Component);
  return _createClass(Giscus, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        repo = _this$props.repo,
        repoId = _this$props.repoId,
        category = _this$props.category,
        categoryId = _this$props.categoryId,
        mapping = _this$props.mapping,
        term = _this$props.term,
        strict = _this$props.strict,
        reactionsEnabled = _this$props.reactionsEnabled,
        emitMetadata = _this$props.emitMetadata,
        inputPosition = _this$props.inputPosition,
        theme = _this$props.theme,
        customThemeCss = _this$props.customThemeCss,
        lang = _this$props.lang,
        lazy = _this$props.lazy;
      if (!repo || !repoId || !categoryId) {
        return (0, _inferno.createVNode)(1, "div", "notification is-danger", [(0, _inferno.createTextVNode)("You forgot to set the "), (0, _inferno.createVNode)(1, "code", null, "repo", 16), (0, _inferno.createTextVNode)(", "), (0, _inferno.createVNode)(1, "code", null, "repoId", 16), (0, _inferno.createTextVNode)(", or "), (0, _inferno.createVNode)(1, "code", null, "categoryId", 16), (0, _inferno.createTextVNode)(' '), (0, _inferno.createTextVNode)("for Giscus. Please set it in "), (0, _inferno.createVNode)(1, "code", null, "_config.yml", 16), (0, _inferno.createTextVNode)(".")], 0);
      }
      if ((mapping === 'specific' || mapping === 'number') && !term) {
        return (0, _inferno.createVNode)(1, "div", "notification is-danger", [(0, _inferno.createTextVNode)("You set "), (0, _inferno.createVNode)(1, "code", null, "mapping", 16), (0, _inferno.createTextVNode)(" to "), (0, _inferno.createVNode)(1, "code", null, "specific", 16), (0, _inferno.createTextVNode)(" or "), (0, _inferno.createVNode)(1, "code", null, "number", 16), (0, _inferno.createTextVNode)(", but did not set "), (0, _inferno.createVNode)(1, "code", null, "term", 16), (0, _inferno.createTextVNode)(" for Giscus. Please set "), (0, _inferno.createVNode)(1, "code", null, "term", 16), (0, _inferno.createTextVNode)(" in "), (0, _inferno.createVNode)(1, "code", null, "_config.yml", 16), (0, _inferno.createTextVNode)(".")], 4);
      }
      var config = {
        repo: repo
      };
      config['data-repo'] = repo;
      config['data-repo-id'] = repoId;
      config['data-category-id'] = categoryId;
      if (category) {
        config['data-category'] = category;
      } else {
        config['data-category'] = 'Announcements';
      }
      if (mapping) {
        config['data-mapping'] = mapping;
      } else {
        config['data-mapping'] = 'pathname';
      }
      if (strict) {
        config['data-strict'] = 1;
      } else {
        config['data-strict'] = 0;
      }
      if (reactionsEnabled) {
        config['data-reactions-enabled'] = 1;
      } else {
        config['data-reactions-enabled'] = 0;
      }
      if (emitMetadata) {
        config['data-emit-metadata'] = 1;
      } else {
        config['data-emit-metadata'] = 0;
      }
      if (inputPosition) {
        config['data-input-position'] = inputPosition;
      } else {
        config['data-input-position'] = 'top';
      }
      if (theme) {
        if (theme === 'custom') {
          if (customThemeCss) {
            config['data-theme'] = customThemeCss;
          } else {
            return (0, _inferno.createVNode)(1, "div", "notification is-danger", [(0, _inferno.createTextVNode)("You set "), (0, _inferno.createVNode)(1, "code", null, "theme", 16), (0, _inferno.createTextVNode)(" to "), (0, _inferno.createVNode)(1, "code", null, "custom", 16), (0, _inferno.createTextVNode)(", but did not apply a"), (0, _inferno.createTextVNode)(' '), (0, _inferno.createVNode)(1, "code", null, "customThemeCss", 16), (0, _inferno.createTextVNode)(" for Giscus. Please set it in "), (0, _inferno.createVNode)(1, "code", null, "_config.yml", 16), (0, _inferno.createTextVNode)(".")], 0);
          }
        } else {
          config['data-theme'] = theme;
        }
      } else {
        config['data-theme'] = 'noborder_light';
      }
      if (lang) {
        config['data-lang'] = lang;
      } else {
        config['data-lang'] = 'en';
      }
      if (lazy) {
        config['data-loading'] = 'lazy';
      }
      return (0, _inferno.normalizeProps)((0, _inferno.createVNode)(1, "script", null, null, 1, _objectSpread(_objectSpread({
        "src": "https://giscus.app/client.js"
      }, config), {}, {
        "crossorigin": "anonymous",
        "async": true
      })));
    }
  }]);
}(Component);
/**
 * Cacheable Giscus comment JSX component.
 * <p>
 * This class is supposed to be used in combination with the <code>locals</code> hexo filter
 * ({@link module:hexo/filter/locals}).
 *
 * @see module:util/cache.cacheComponent
 * @example
 * <Giscus.Cacheable
 *     comment={{
 *         repo: "usr/repo"
 *         repoId: "X_xxxxxxxxxx"
 *         category: "******"
 *         categoryId: "XXX_xxxxxxxxxxxxxxxx"
 *         mapping: "******"
 *         strict: {false}
 *         reactionsEnabled: {false}
 *         emitMetadata: {false}
 *         inputPosition: "******"
 *         theme: "******"
 *         lang: "******"
 *         lazy: "******"
 *     }} />
 */
Giscus.Cacheable = cacheComponent(Giscus, 'comment.giscus', function (props) {
  var _props$comment = props.comment,
    repo = _props$comment.repo,
    repoId = _props$comment.repoId,
    category = _props$comment.category,
    categoryId = _props$comment.categoryId,
    mapping = _props$comment.mapping,
    term = _props$comment.term,
    strict = _props$comment.strict,
    reactionsEnabled = _props$comment.reactionsEnabled,
    emitMetadata = _props$comment.emitMetadata,
    inputPosition = _props$comment.inputPosition,
    theme = _props$comment.theme,
    customThemeCss = _props$comment.customThemeCss,
    lang = _props$comment.lang,
    lazy = _props$comment.lazy;
  return {
    repo: repo,
    repoId: repoId,
    category: category,
    categoryId: categoryId,
    mapping: mapping,
    term: term,
    strict: strict,
    reactionsEnabled: reactionsEnabled,
    emitMetadata: emitMetadata,
    inputPosition: inputPosition,
    theme: theme,
    customThemeCss: customThemeCss,
    lang: lang,
    lazy: lazy
  };
});
module.exports = Giscus;
